materialize-boilerplate/apply.go

```diff
@@ -0,0 +1,229 @@
+	// the destination. It's called for every binding, although it may not have any `BindingUpdate`
+	// parameters. This is to allow materializations to perform additional specific actions on
+	// binding changes that are not covered by the general cases of the `BindingUpdate` parameters.
+	UpdateResource(ctx context.Context, spec *pf.MaterializationSpec, bindingIndex int, applyParams BindingUpdate) (string, ActionApplyFn, error)
+}
+
+// ApplyChanges applies changes to an endpoint. It computes these changes from the apply request and
+// the state of the endpoint per the `*InfoSchema`. The `Applier` executes the resulting actions,
+// optionally with a concurrent scatter/gather for expedience on endpoints that would benefit from
+// that sort of thing.
+func ApplyChanges(ctx context.Context, req *pm.Request_Apply, applier Applier, is *InfoSchema, concurrent bool) (*pm.Response_Applied, error) {
+	if err := req.Validate(); err != nil {
+		return nil, fmt.Errorf("validating request: %w", err)
+	}
+
+	storedSpec, err := applier.LoadSpec(ctx, req.Materialization.Name)
+	if err != nil {
+		return nil, fmt.Errorf("getting stored spec: %w", err)
+	}
+
+	actionDescriptions := []string{}
+	actions := []ActionApplyFn{}
+
+	addAction := func(desc string, a ActionApplyFn) {
+		if a != nil { // Convenience for handling endpoints that return `nil` for a no-op action.
+			actionDescriptions = append(actionDescriptions, desc)
+			actions = append(actions, a)
+		}
+	}
+
+	// Only attempt to create meta tables if the storedSpec was `nil`, which suggests that the meta
+	// tables don't yet exist.
+	// TODO(whb): We will eventually stop persisting specs for materializations, and instead include
+	// the previous spec as part of the protocol. When that happens this will need updated.
+	if storedSpec == nil {
+		desc, action, err := applier.CreateMetaTables(ctx, req.Materialization)
+		if err != nil {
+			return nil, fmt.Errorf("getting CreateMetaTables action: %w", err)
+		}
+		addAction(desc, action)
+	}
+
+	for bindingIdx, binding := range req.Materialization.Bindings {
+		// The existing binding spec is used to extract various properties that can't be learned
+		// from introspecting the destination system, such as the backfill counter and if the
+		// materialization was previously delta updates.
+		existingBinding, err := FindExistingBinding(binding.ResourcePath, binding.Collection.Name, storedSpec)
+		if err != nil {
+			return nil, fmt.Errorf("finding existing binding: %w", err)
+		}
+
+		if !is.HasResource(binding.ResourcePath) {
+			// Resource does not yet exist, and must be created.
+			desc, action, err := applier.CreateResource(ctx, req.Materialization, bindingIdx)
+			if err != nil {
+				return nil, fmt.Errorf("getting CreateResource action: %w", err)
+			}
+			addAction(desc, action)
+		} else if existingBinding != nil && existingBinding.Backfill != binding.Backfill {
```

Comment:
What about a DropResource action of the Applier interface, and remove ReplaceResource ?

That would make it an easier change, if we decide to tweak the circumstances under which we drop tables.

Reply:
I think you're alluding to make a table replacement always be a drop followed by a create, which is an interesting idea since that's what some systems have to do anyway.

We've already got the implementations for all materialization done and working with a distinct "replace" action, which will sometimes do an actual "replace" if the destination supports it, sometimes do a drop & create within a transaction, and failing all of that does a non-transactional drop & replace. So I'm not too inclined to rework all of those right now, and maybe this could be a future improvement if we need to tweak dropping tables.

Reply:
:+1: to doing this in the future, but I continue to think separate Drop and Create callbacks are more appropriate. The current Replace feels like an over-generalization of the CREATE OR REPLACE sql pattern... which we just saw today (with bigquery clustering keys) doesn't always work either.